src/thor/timedep_forward.cc

```diff
@@ -149,17 +148,40 @@ inline bool TimeDepForward::ExpandForwardInner(GraphReader& graphreader,
                                                const valhalla::Location& destination,
                                                std::pair<int32_t, float>& best_path) {
 
-  // Skip shortcut edges for time dependent routes. Also skip this edge if permanently labeled
-  // (best path already found to this directed edge), if no access is allowed to this edge
+  // Skip shortcut edges for time dependent routes, if no access is allowed to this edge
   // (based on costing method), or if a complex restriction exists.
   bool has_time_restrictions = false;
-  if (meta.edge->is_shortcut() || meta.edge_status->set() == EdgeSet::kPermanent ||
-      !costing_->Allowed(meta.edge, pred, tile, meta.edge_id, localtime, nodeinfo->timezone(),
-                         has_time_restrictions) ||
-      costing_->Restricted(meta.edge, pred, edgelabels_, tile, meta.edge_id, true, localtime,
-                           nodeinfo->timezone())) {
+  if (meta.edge->is_shortcut() || !costing_->Allowed(meta.edge, pred, tile, meta.edge_id, localtime,
+                                                     nodeinfo->timezone(), has_time_restrictions)) {
     return false;
   }
+  bool is_complex_restriction = costing_->Restricted(meta.edge, pred, edgelabels_, tile, meta.edge_id,
+                                                     true, localtime, nodeinfo->timezone());
+
+  if (is_complex_restriction) {
+    // A complex restriction spans multiple edges, e.g. from A to C via B
+    // At this point we've evaluated C and know that the restriction is active
+    // in our circumstances. However, B is still marked as kPermanent since
+    // we didn't know at that time that A to B would eventually form a restricted path
+    //
+    // So, in order to still allow new paths involving B, e.g. D to B to C,
+    // we need to go back and revert the permanent status of B
+    //
+    // We mark it as kUnreached so that in effect it is no longer visible. (It can't
+    // be removed since that invalidates subsequent indices and setting it to temporary
+    // means it'll still do a comparison to existing sort cost and fail later).
+    //
+    // TODO can complex restrictions be arbitrarily long? Is so, walk the graph and
+    // reset everything but the first
+    // TODO can pred be the first edge in complex restriction? If so, this fails
+    edgestatus_.Update(pred.edgeid(), EdgeSet::kUnreached);
```

Comment:
I've been going over the `adjacency_list_` and the `edgestatus_` but with all logic around these containers, I'm not fully sure that that this change to kUnreached here works in all cases. Would like some extra eyes here for the review.

Reply:
Yeah my first thought is exactly what your TODO says. I think we would have to walk back the edges in the restriction and unmark all of them but the first. Yes a restriction can be longer than just 3 edges.

Reply:
Pushed new code that walks the restriction.

Reply:
Also, I ended up resetting _all_ of the edges in the restriction. You can imagine a case where a restriction for A->B->C still has a best valid route that goes through A, so I think _all_ must be reset, including A.

Reply:
@danpaz raised a good question: Would it be possible to enter an infinite loop of expanding the same edges forever?

Reply:
After some more thinking, I believe it _would_ be possible to enter an infinite loop if there are at least two overlapping complex restrictions biting each others tails.

I therefore went back to the original design of resetting all edges in the restriction _excluding_ the first edge in the restriction: https://github.com/valhalla/valhalla/pull/2109/commits/b49c1b2ba11a6581e0102f1c4777688ecbf3b976

My earlier statement of:
> You can imagine a case where a restriction for A->B->C still has a best valid route that goes through A, so I think all must be reset, including A.

I believe to be false since the imagined case _should_ still be found, regardless of if A was reset or not.

Reply:
Not sure I'm following the entire discussion here, but un-labeling kPermanent I don't think solves it. Yes, it can the revisit the edge but each edge can only store one predecessor (the one with the lowest cost to it).

Reply:
@dnesbitt61 I added a testcase in https://github.com/valhalla/valhalla/pull/2109/commits/ee89f9d74b891a05eb6372276bb0a6d123fff972 which I believe tests the regression you brought up.

Essentially, it tests if a path is found that uses part of a complex restriction, after the full restriction triggered a reset of the EdgeStatus to `kUnreached`. The next commit renames `kUnreached` to `kUnreachedOrReset` to better convey this process.

Reply:
I think a couple test cases are needed: 1) where the destination must take the complex turn restriction - should take the longer path to get to the node where the restriction occurs, 2) where the destination does not take the restricted turn - should take the short path to get to the node rather than the longer path. I am afraid that solving for case 1 will create bad paths for case 2 since there can only be one shortest path to the end node.